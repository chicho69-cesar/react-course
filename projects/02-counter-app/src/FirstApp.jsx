import PropTypes from 'prop-types'
import { Fragment } from 'react'

export const FirstApp = ({ title, subtitle, name }) => {
  return (
    <Fragment>
      <h1 data-testid="test-title">{title}</h1>
      <p>{subtitle}</p>
      <p>{name}</p>
    </Fragment>
  )
}

FirstApp.propTypes = {
  title: PropTypes.string.isRequired,
  subtitle: PropTypes.string,
  name: PropTypes.string
}
