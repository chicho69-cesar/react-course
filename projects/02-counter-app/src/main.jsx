import './index.css'

import { StrictMode } from 'react'
import { createRoot } from 'react-dom/client'
import { CounterApp } from './CounterApp'

// import { HelloWorldApp } from './HelloWorldApp.jsx'
// import { FirstApp } from './FirstApp'

const rootElement = document.getElementById('root')
const root = createRoot(rootElement)

root.render(
  <StrictMode>
    {/* <HelloWorldApp /> */}
    
    {/* <FirstApp
      title={'Hola, soy Goku'}
      subtitle={'Soy un subtÃ­tulo'}
      name={'Goku'}
    /> */}

    <CounterApp value={10} />
  </StrictMode>,
)
