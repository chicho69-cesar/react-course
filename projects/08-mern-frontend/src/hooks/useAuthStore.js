import { useDispatch, useSelector } from 'react-redux'

import calendarApi from '../api/calendarApi'
import { clearErrorMessage, onChecking, onLogin, onLogout } from '../store/auth/authSlice'
import { onLogoutCalendar } from '../store/calendar/calendarSlice'

export const useAuthStore = () => {
  const { status, user, errorMessage } = useSelector((state) => state.auth)
  const dispatch = useDispatch()

  const startLogin = async ({ email, password }) => {
    dispatch(onChecking())

    try {
      const { data } = await calendarApi.post('/auth/login', { email, password })

      localStorage.setItem('token', data.token)
      localStorage.setItem('token-init-date', new Date().getTime())

      dispatch(onLogin({ name: data.name, uid: data.uid }))
    } catch (error) {
      console.error('Error during login:', error)
      dispatch(onLogout('Credenciales incorrectas'))

      setTimeout(() => {
        dispatch(clearErrorMessage())
      }, 100)
    }
  }

  const startRegister = async ({ email, password, name }) => {
    dispatch(onChecking())

    try {
      const { data } = await calendarApi.post('/auth/register', { email, password, name })

      localStorage.setItem('token', data.token)
      localStorage.setItem('token-init-date', new Date().getTime())

      dispatch(onLogin({ name: data.name, uid: data.uid }))
    } catch (error) {
      console.error('Error during login:', error)
      dispatch(onLogout(error.response.data?.msg || 'Error al registrar usuario'))

      setTimeout(() => {
        dispatch(clearErrorMessage())
      }, 100)
    }
  }

  const checkAuthToken = async () => {
    const token = localStorage.getItem('token')
    if (!token) return dispatch(onLogout())

    try {
      const { data } = await calendarApi.get('/auth/renew')

      localStorage.setItem('token', data.token)
      localStorage.setItem('token-init-date', new Date().getTime())

      dispatch(onLogin({ name: data.name, uid: data.uid }))
    } catch (error) {
      console.error('Error checking auth token:', error)

      localStorage.clear()
      dispatch(onLogout())

      return
    }
  }

  const startLogout = () => {
    localStorage.clear()
    dispatch(onLogoutCalendar())
    dispatch(onLogout())
  }

  return {
    status,
    user,
    errorMessage,

    startLogin,
    startRegister,
    checkAuthToken,
    startLogout,
  }
}
