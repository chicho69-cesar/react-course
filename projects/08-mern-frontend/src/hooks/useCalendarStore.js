import { useDispatch, useSelector } from 'react-redux'
import Swal from 'sweetalert2'

import calendarApi from '../api/calendarApi'
import { convertEventsToDateEvents } from '../helpers'
import { onAddNewEvent, onDeleteEvent, onLoadEvents, onSetActiveEvent, onSetIsForUpdate, onUpdateEvent } from '../store/calendar/calendarSlice'

export const useCalendarStore = () => {
  const dispatch = useDispatch()
  const { events, activeEvent, isForUpdate } = useSelector((state) => state.calendar)
  const { user } = useSelector((state) => state.auth)

  const setActiveEvent = (calendarEvent) => {
    dispatch(onSetActiveEvent(calendarEvent))
  }
  
  const setIsForUpdate = (isForUpdate) => {
    dispatch(onSetIsForUpdate(isForUpdate))
  }

  const startLoadingEvents = async () => {
    try {
      const { data } = await calendarApi.get('/events')
      const events = convertEventsToDateEvents(data.events)
      dispatch(onLoadEvents(events))
    } catch (error) {
      console.error('Error loading events:', error)
      Swal.fire('Error al cargar eventos', 'Hubo un error al cargar los eventos', 'error')
    }
  }

  const startSavingEvent = async (calendarEvent) => {
    try {
      if (calendarEvent.id) {
        await calendarApi.put(`/events/${calendarEvent.id}`, calendarEvent)
        dispatch(onUpdateEvent({ ...calendarEvent, user }))

        return
      }

      const { data } = await calendarApi.post('/events', calendarEvent)
      dispatch(onAddNewEvent({ ...calendarEvent, id: data.event.id, user }))
    } catch (error) {
      console.error('Error saving event:', error)
      Swal.fire('Error al guardar', 'Hubo un error al guardar el evento', 'error')
    }
  }

  const startDeletingEvent = async () => {
    try {
      await calendarApi.delete(`/events/${activeEvent.id}`)
      dispatch(onDeleteEvent())
    } catch (error) {
      console.error('Error deleting event:', error)
      Swal.fire('Error al eliminar', 'Hubo un error al eliminar el evento', 'error')
    }
  }

  return {
    events,
    activeEvent,
    isForUpdate,
    hasEventSelected: !!activeEvent,
    setActiveEvent,
    setIsForUpdate,
    startLoadingEvents,
    startSavingEvent,
    startDeletingEvent
  }
}
