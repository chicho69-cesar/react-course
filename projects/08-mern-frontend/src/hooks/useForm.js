import { useEffect, useMemo, useState } from 'react'

export const useForm = (initialForm = {}, formValidations = {}) => {
  const [formState, setFormState] = useState(initialForm)
  const [formValidationState, setFormValidationState] = useState({})

  useEffect(() => {
    createValidators()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [formState])

  useEffect(() => {
    setFormState(initialForm)
  }, [initialForm])

  const isFormValid = useMemo(() => {
    for (const formValue of Object.keys(formValidationState)) {
      if (formValidationState[formValue] !== null) return false
    }

    return true
  }, [formValidationState])

  const onInputChange = ({ target }) => {
    const { name, value } = target

    setFormState({
      ...formState,
      [name]: value
    })
  }

  const onResetForm = () => {
    setFormState(initialForm)
    setFormValidationState({})
  }

  const createValidators = () => {
    const formCheckedValues = {}

    for (const formField of Object.keys(formValidations)) {
      const [fn, errorMessage] = formValidations[formField]
      formCheckedValues[`${formField}Valid`] = fn(formState[formField]) ? null : errorMessage
    }

    setFormValidationState(formCheckedValues)
  }

  return {
    ...formState,
    formState,
    onInputChange,
    onResetForm,

    ...formValidationState,
    isFormValid,
    formValidationState,
  }
}
