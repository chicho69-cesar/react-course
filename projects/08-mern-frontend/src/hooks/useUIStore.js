import { useDispatch, useSelector } from 'react-redux'
import { onOpenDateModal, onCloseDateModal } from '../store/ui/uiSlice'

export const useUIStore = () => {
  const { isDateModalOpen } = useSelector((state) => state.ui)
  const dispatch = useDispatch()

  const openDateModal = () => {
    dispatch(onOpenDateModal())
  }

  const closeDateModal = () => {
    dispatch(onCloseDateModal())
  }

  const toggleDateModal = () => {
    isDateModalOpen ? closeDateModal() : openDateModal()
  }

  return {
    isDateModalOpen,
    openDateModal,
    closeDateModal,
    toggleDateModal
  }
}
