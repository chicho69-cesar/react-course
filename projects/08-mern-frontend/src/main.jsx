import './index.css'

// import { StrictMode } from 'react'
import { createRoot } from 'react-dom/client'

import App from './App'

const rootItem = document.querySelector('#root')
const root = createRoot(rootItem)

root.render(
  // <StrictMode>
  <App />
  // </StrictMode>
)
