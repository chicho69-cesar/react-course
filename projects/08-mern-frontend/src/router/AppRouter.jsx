import { useEffect } from 'react'
import { Navigate, Route, Routes } from 'react-router'

import { AuthPage } from '../auth/pages/AuthPage'
import { CalendarPage } from '../calendar'
import { useAuthStore } from '../hooks'

export const AppRouter = () => {
  const { status, checkAuthToken } = useAuthStore()

  useEffect(() => {
    checkAuthToken()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  if (status === 'checking') {
    return (
      <div className='loading-screen'>
        <h2>Loading...</h2>
      </div>
    )
  }

  return (
    <Routes>
      {status === 'not-authenticated' ? (
        <>
          <Route path='/auth/*' element={<AuthPage />} />
          <Route path='/*' element={<Navigate to='/auth/login' />} />
        </>
      ) : (
        <>
          <Route path='/' element={<CalendarPage />} />
          <Route path='/*' element={<Navigate to='/' />} />
        </>
      )}
    </Routes>
  )
}
