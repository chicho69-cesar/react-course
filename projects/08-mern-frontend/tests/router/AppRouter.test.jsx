/* eslint-disable no-undef */
import { render, screen } from '@testing-library/react'
import { MemoryRouter } from 'react-router'

import { useAuthStore } from '../../src/hooks'
import { AppRouter } from '../../src/router'

jest.mock('../../src/hooks/useAuthStore')

jest.mock('../../src/calendar', () => ({
  CalendarPage: () => <h1>CalendarPage</h1>
}))

describe('Tests on <AppRouter />', () => {
  const mockCheckAuthToken = jest.fn()

  beforeEach(() => {
    jest.clearAllMocks()
  })

  test('Should show loading screen and call check auth token', () => {
    useAuthStore.mockReturnValue({
      status: 'checking',
      checkAuthToken: mockCheckAuthToken
    })

    render(
      <AppRouter />
    )

    expect(screen.getByText('Loading...')).toBeTruthy()
    expect(mockCheckAuthToken).toHaveBeenCalled()
  })

  test('Should show auth page if it\'s not authenticated', () => {
    useAuthStore.mockReturnValue({
      status: 'not-authenticated',
      checkAuthToken: mockCheckAuthToken
    })

    const { container } = render(
      <MemoryRouter initialEntries={['/auth/test/other-screen']}>
        <AppRouter />
      </MemoryRouter>
    )

    expect(screen.getAllByText('Iniciar Sesi√≥n').length).toBeGreaterThan(0)
    expect(container).toMatchSnapshot()
  })
})
