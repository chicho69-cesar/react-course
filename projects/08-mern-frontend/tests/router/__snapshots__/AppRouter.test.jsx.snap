// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <AppRouter /> Should show auth page if it's not authenticated 1`] = `
<div
  aria-hidden="true"
>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <section
        class="col-md-6 login-form-1"
      >
        <h3>
          Iniciar Sesión
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginEmail"
              placeholder="Correo"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginPassword"
              placeholder="Contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2"
          >
            <button
              class="btnSubmit"
              type="submit"
            >
              Iniciar Sesión
            </button>
          </div>
        </form>
      </section>
      <section
        class="col-md-6 login-form-2"
      >
        <h3>
          Registrarse
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerName"
              placeholder="Nombre"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerEmail"
              placeholder="Correo"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword"
              placeholder="Contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerConfirmPassword"
              placeholder="Confirmar Contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2"
          >
            <button
              class="btnSubmit"
              type="submit"
            >
              Registrarse
            </button>
          </div>
        </form>
      </section>
    </div>
  </div>
</div>
`;
